runner_test.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+// Use `go test -tree` to debug DumpRunner tests.
+var printTree = flag.Bool("tree", false, "print file tree in DumpRunner tests")
+
+func TestDumpRunner(t *testing.T) {
+	dir, err := ioutil.TempDir("", "test-dumprunner-dir-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	if *printTree {
+		defer func() { t.Log("\n" + tree(dir)) }()
+	}
+
+	// We will use dr several times to make sure it does what it should do.
+	// Each subtest begins with setting the cmd we will run.
+	dr := NewDumpRunner(dir)
+
+	// #1 -- common case, command exits with 0, output is saved to a file.
+	cmd := helperCommand("echo", "ok")
+
+	// Set cmd.Stdout to later ensure that dr.Run will write to both b and a
+	// file.
+	var b bytes.Buffer
+	cmd.Stdout = &b
+
+	// These are the files and contents we expect to see:
+	names := []string{"out"}
+	contents := []string{cmd.Args[len(cmd.Args)-1] + "\n"}
+
+	if err := dr.Run(cmd, names[len(names)-1]); err != nil {
+		t.Errorf("Run(%q) = %v, want %v", cmd.Args[3:], err, nil)
+	}
+	if err := dirHasExactFiles(dir, names, contents); err != nil {
+		t.Error(err)
+	}
+	if got, want := b.String(), contents[len(contents)-1]; got != want {
```

Comment:
this is cool, I didn't know you could do multiple comparisons like this, I assume this is compared with `and`? In other words, it's the same as writing this:
```
if got == b.String() && wants == contents[len(contents)-1]{
```